src/transport/tcp/connection.rs

```diff
@@ -455,9 +459,21 @@ impl TcpConnection {
             AddressType::Socket(address) => Multiaddr::empty()
                 .with(Protocol::from(address.ip()))
                 .with(Protocol::Tcp(address.port())),
-            AddressType::Dns(address, port) => Multiaddr::empty()
-                .with(Protocol::Dns(Cow::Owned(address)))
-                .with(Protocol::Tcp(port)),
+            AddressType::Dns {
+                address,
+                port,
+                dns_type,
+            } => match dns_type {
+                DnsType::Dns => Multiaddr::empty()
+                    .with(Protocol::Dns(Cow::Owned(address)))
+                    .with(Protocol::Tcp(port)),
+                DnsType::Dns4 => Multiaddr::empty()
+                    .with(Protocol::Dns4(Cow::Owned(address)))
+                    .with(Protocol::Tcp(port)),
+                DnsType::Dns6 => Multiaddr::empty()
+                    .with(Protocol::Dns6(Cow::Owned(address)))
+                    .with(Protocol::Tcp(port)),
+            },
```

Comment:
Here we had another miss-match, where regardless of the initial DNS version we'd accept both ipv4 or ipv6